### PR TITLE
fix deprecation warning when wcs is initialized with a pipeline 

### LINKED ISSUE #361: DeprecationWarning: Indexing a WCS.pipeline step is deprecated is raised when a WCS is initialized with a pipeline
We are seeing the following deprecation warning in `jwst`:

```
$ pytest jwst/assign_mtwcs/ -W "error: Indexing"
============================ test session starts ============================
platform darwin -- Python 3.9.2, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
crds_context: jwst_0712.pmap
rootdir: /Users/jdavies/dev/jwst, configfile: setup.cfg
plugins: requests-mock-1.8.0, cov-2.11.1, ci-watson-0.5, asdf-2.7.3, xdist-2.2.1, openfiles-0.5.0, forked-1.3.0, doctestplus-0.9.0
collected 1 item                                                            

jwst/assign_mtwcs/tests/test_mtwcs.py F                               [100%]

================================= FAILURES ==================================
_____________________________ test_mt_multislit _____________________________

    def test_mt_multislit():
        file_path = os.path.join(data.__path__[0], 'test_mt_asn.json')
        with datamodels.open(file_path) as model:
            assert model[0].slits[0].meta.wcs.output_frame.name == 'world'
            step = AssignMTWcsStep()
>           result = step.run(model)

jwst/assign_mtwcs/tests/test_mtwcs.py:15: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../miniconda3/envs/jwst/lib/python3.9/site-packages/stpipe/step.py:407: in run
    step_result = self.process(*args)
jwst/assign_mtwcs/assign_mtwcs_step.py:40: in process
    result = assign_moving_target_wcs(input)
jwst/assign_mtwcs/moving_target_wcs.py:49: in assign_moving_target_wcs
    new_wcs = add_mt_frame(slit.meta.wcs,
jwst/assign_mtwcs/moving_target_wcs.py:103: in add_mt_frame
    new_wcs = WCS(pipeline)
../../miniconda3/envs/jwst/lib/python3.9/site-packages/gwcs/wcs.py:139: in __init__
    self._initialize_wcs(forward_transform, input_frame, output_frame)
../../miniconda3/envs/jwst/lib/python3.9/site-packages/gwcs/wcs.py:159: in _initialize_wcs
    name, frame_obj = self._get_frame_name(item[0])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = Step(frame=detector,                       transform=dms2sca), ind = 0

    def __getitem__(self, ind):
>       warnings.warn("Indexing a WCS.pipeline step is deprecated. "
                      "Use the `frame` and `transform` attributes instead.", DeprecationWarning)
E       DeprecationWarning: Indexing a WCS.pipeline step is deprecated. Use the `frame` and `transform` attributes instead.

../../miniconda3/envs/jwst/lib/python3.9/site-packages/gwcs/wcs.py:2833: DeprecationWarning
========================== short test summary info ==========================
FAILED jwst/assign_mtwcs/tests/test_mtwcs.py::test_mt_multislit - Deprecat...
============================= 1 failed in 3.88s =============================
```

It doesn't look like one can initialize a WCS with a list of `(frame, transform)` tuples without triggering this `DeprecationWarning`.

### PR DESCRIPTION
**Pull request by cshanahan1**
It wasn't a problem with initializing a wcs with a list of tuples of `(frame, transform)` - it happened because there were `wcs.Step()` items in the `forward_transform` list the wcs was being initialized with which gave that warning when trying to index them rather than accessing their `.frame` and `.transform` attributes.

Fixes #361

### CODE REVIEW

**Review by jdavies-st — APPROVED** [review]
Looks good.  Fixes the warning we're seeing in `jwst`.

You'll want to add a small test that runs the code through the `else` blocks and does something like:

```python
with pytest.warns(DeprecationWarning, match="Indexing a WCS.pipeline step is deprecated"):
    do stuff that goes through the else block
```

or similar.

**Comment by cshanahan1** [comment]
It wasn't a problem with initializing a wcs with a list of tuples of (frame, transform) - it happened because there were wcs.Step() items in the forward_transform list the wcs was being initialized with which gave that warning when trying to index them rather than accessing their .frame and .transform attributes. 

I also used `pytest.mark.filterwarnings` to declare that particular warning as an error and fail the test if it appears, it looks like the pytest.warns example above is useful to ensure a warning DOES occur?

**Comment by jdavies-st** [comment]
Looks great! And thanks for the clarification.  I added that text to the top of the PR.  Just add an entry to CHANGES.rst and we can merge.

File: gwcs/tests/test_wcs.py [hunk 1]
```diff
@@ -1056,3 +1056,19 @@ def test_tabular_2d_quantity():
...
+def test_initialize_wcs_with_list():
+    # test that you can  initialize a wcs with a pipeline that is a list
+    # containing both Step() and (frame, transform) tuples
+
+    trans = models.Shift(10 * u .pix) & models.Shift(2 * u.pix)
+    pipeline = [('detector', trans), ('sky', None)]
+
+    # add a Step() to the pipeline list
+    extra_step = wcs.Step('extra_step', trans)
+    pipeline.append(extra_step)
```
Comment (nden) [inline gwcs/tests/test_wcs.py:1071]: This test creates the pipeline incorrectly. The `None` sentinel must be in the last step. The errors is currently raised only when the WCS is printed or executed. There should be a check in the initialization of the WCS.

File: gwcs/wcs.py [hunk 2]
```diff
@@ -138,7 +138,14 @@ def __init__(self, forward_transform=None, input_frame='detector', output_frame=
...
         self._pixel_shape = None
-        self._pipeline = [Step(*step) for step in self._pipeline]
+
+        pipe = []
+        for step in self._pipeline:
+            if isinstance(step, Step):
+                pipe.append(Step(step.frame, step.transform))
+            else:
+                pipe.append(Step(*step))
+        self._pipeline = pipe
```
Comment (nden) [inline gwcs/wcs.py:148]: This code may be moved to the `_initialize_wcs` function.

**Review by nden — CHANGES_REQUESTED** [review]

**Comment by nden** [comment]
It doesn't solve the issue completely, gwcs tests (specifically test_get_transfrom) still raise the warning.
I think it's the right approach but needs more work.

**Comment by cshanahan1** [comment]
@nden the reason that test_get_transform was raising the warning was because it was testing that indexing a pipeline step works the same as accessing the transform attribute
`    assert_allclose(w.pipeline[0].transform(x, y), (fx, fy))`
`    assert_allclose(w.pipeline[0][1](x, y), (fx, fy))`

If it's actually being deprecated, should that second line just be removed from the tests? Or is the issue that indexing a step shouldn't raise a deprecation warning at all anymore, because right now that warning is in getitem for Step.

**Comment by cshanahan1** [comment]
And similarly, the tests in test_init_no_transform() are testing because they're comparing 

` pipeline[0][0]`
to
`pipeline[0].frame` 

etc, which is obviously raising that warning because that's trying to index a step. The fix I put here was indexing of steps that was happening internally which is not fixed, whereas this is indexing that is explicitly written in a test.

**Comment by nden** [comment]
Needs a rebase and please fix a code style issue [here](https://github.com/spacetelescope/gwcs/runs/3452321810?check_suite_focus=true). I merged the other PR with the error.

File: gwcs/tests/test_wcs.py [hunk 3]
```diff
@@ -1056,3 +1056,19 @@ def test_tabular_2d_quantity():
...
     assert all(u.allclose(u.Quantity(b), [0, 2] * u.pix) for b in bb)
+
+
+@pytest.mark.filterwarnings("error:.*Indexing a WCS.pipeline step is deprecated.*:DeprecationWarning")
+def test_initialize_wcs_with_list():
+    # test that you can  initialize a wcs with a pipeline that is a list
+    # containing both Step() and (frame, transform) tuples
+
+    trans = models.Shift(10 * u .pix) & models.Shift(2 * u.pix)
+    pipeline = [('detector', trans), ('sky', None)]
```
Comment (nden) [inline gwcs/tests/test_wcs.py:1067]: Only the last step in a pipeline can have a transform `None`.

**Comment by jdavies-st** [comment]
> Needs a rebase and please fix a code style issue [here](https://github.com/spacetelescope/gwcs/runs/3452321810?check_suite_focus=true). I merged the other PR with the error.

I fixed the code style error in #372.

**Review by nden — APPROVED** [review]

**Pull request merged**